### PR TITLE
[WIP] Reject impl keyword inside impl header

### LINKED ISSUE #77: Improve parser recovery for impls
Currently, the recovery for 

```
impl<T: Clone>
impl<T> OnceCell<T> {
}
```

Is bad (see https://github.com/matklad/rust-analyzer/commit/d76c3dd8fc2960c044e37fa5503383c15344d1a5, we want the second impl to be parsed as an impl).

The problem is that we try to parse a type after and `impl`, and `impl` keyword can begin an impl trait type, so we interpret the second impl as a start of a type, instead of bailing out. I think we shouldn't try to parse `impl trait` type in this position. 

I thought that [original parser](https://github.com/rust-lang/rust/blob/4f3ff5a97bcd2d05ee0c768122752dc74f96ccc3/src/libsyntax/parse/parser.rs#L5837) also specifically parsed only a subset of types in this position, but it doesn't seem to be the case?

The relevant code is here:

https://github.com/matklad/rust-analyzer/blob/f5808b21a423235037a487d877e23deb3e602f37/crates/ra_syntax/src/grammar/items/traits.rs#L58

I think we should pass some flag to type to say "parse only stuff that can appear in impl header"

### PR DESCRIPTION
**Pull request by csmoe**
Closes #77

### CODE REVIEW

**Comment by csmoe** [comment]
oops, the approach is very incorrect.

**Comment by csmoe** [comment]
@matklad In the origin parser `impl Trait` inside impl header is [parsed as type](https://play.rust-lang.org/?gist=8576ba550f322ba3a49a5240c2087dc6&version=stable&mode=debug&edition=2015):
```rust
error: expected a trait, found type
 --> src/main.rs:2:10
  |
2 |     impl impl Send for i32 { }
  |          ^^^^^^^^^

error: aborting due to previous error
```
compare to sub-impl-header:
```rust
error: expected identifier, found `<`
 --> src/main.rs:2:14
  |
2 |     impl impl<T> Send for T { }
  |              ^ expected identifier

error: aborting due to previous error
```
so should I treat them separately instead of simply rejecting them at once?

**Comment by matklad** [comment]
I think the approach here is sound: just bail early if you see an `impl`. However, I think a slightly better implementation would to introduce a helper around `type_`

```
fn impl_type(p: &mut Parser) {
    if p.at(IMPL_KEYWORD) { p.error(); return } 
    type_(p);
}
```

That is, we *should* be different from the upstream parser I think.

File: crates/ra_syntax/tests/data/parser/err/0026_imp_recovery.rs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+impl<T: Clone>
+impl<T:Clone>
+!impl<T> OnceCell<T> {
```
Comment (matklad) [inline crates/ra_syntax/tests/data/parser/err/0026_imp_recovery.rs]: I think we don't need a `!` in the test.

File: crates/ra_syntax/src/grammar/items/traits.rs [hunk 2]
```diff
@@ -115,3 +115,17 @@ fn choose_type_params_over_qpath(p: &Parser) -> bool {
     (p.nth(1) == LIFETIME || p.nth(1) == IDENT)
         && (p.nth(2) == R_ANGLE || p.nth(2) == COMMA || p.nth(2) == COLON || p.nth(2) == EQ)
 }
+
+// impl Type {}
+//      ^^^^
+// impl Trait for T {}
+//      ^^^^^
+pub(crate) fn impl_type(p: &mut Parser) {
```
Comment (matklad) [inline crates/ra_syntax/src/grammar/items/traits.rs]: You can turn these comments into executable tests, by changing them to 

```
// test impl_type
// impl Trait for Type {}
// impl impl NotType {}
// impl Trait for impl NotType {}
```

, running `cargo gen-tests` to generate tests at tests/data/parser/inline and running `cargo test` to generate the gold answer.

**Comment by matklad** [comment]
bors r+

Thanks!

**Pull request merged**